File: integrations/java/app/revanced/integrations/patches/PlayerOverlaysHookPatch.java
package app.revanced.integrations.patches;

import android.view.ViewGroup;

import androidx.annotation.Nullable;

import app.revanced.integrations.shared.PlayerOverlays;

/**
 * Hook receiver class for 'player-overlays-hook' patch
 *
 * @usedBy app.revanced.patches.youtube.misc.playeroverlay.patch.PlayerOverlaysHookPatch
 * @smali Lapp/revanced/integrations/patches/PlayerOverlaysHookPatch;
 */
@SuppressWarnings("unused")
public class PlayerOverlaysHookPatch {
    /**
     * Hook into YouTubePlayerOverlaysLayout.onFinishInflate() method
     *
     * @param thisRef reference to the view
     * @smali YouTubePlayerOverlaysLayout_onFinishInflateHook(Ljava / lang / Object ;)V
     */
    public static void YouTubePlayerOverlaysLayout_onFinishInflateHook(@Nullable Object thisRef) {
        if (thisRef == null) return;
        if (thisRef instanceof ViewGroup) {
            PlayerOverlays.attach((ViewGroup) thisRef);
        }
    }
}

File: src/main/kotlin/app/revanced/patches/youtube/misc/playeroverlay/PlayerOverlaysHookPatch.kt
package app.revanced.patches.youtube.misc.playeroverlay

import app.revanced.patcher.data.BytecodeContext
import app.revanced.patcher.extensions.InstructionExtensions.addInstruction
import app.revanced.patcher.patch.BytecodePatch
import app.revanced.patcher.patch.annotation.Patch
import app.revanced.patches.youtube.misc.integrations.IntegrationsPatch
import app.revanced.patches.youtube.misc.playeroverlay.fingerprint.PlayerOverlaysOnFinishInflateFingerprint
import app.revanced.util.exception

@Patch(
    description = "Hook for the video player overlay",
    dependencies = [IntegrationsPatch::class],
)

/**
 * Edit: This patch is not in use and may not work.
 */
@Suppress("unused")
object PlayerOverlaysHookPatch : BytecodePatch(
    setOf(PlayerOverlaysOnFinishInflateFingerprint)
) {
    private const val INTEGRATIONS_CLASS_DESCRIPTOR = "Lapp/revanced/integrations/youtube/patches/PlayerOverlaysHookPatch;"

    override fun execute(context: BytecodeContext) {
        PlayerOverlaysOnFinishInflateFingerprint.result?.mutableMethod?.apply {
            addInstruction(
                implementation!!.instructions.lastIndex,
                "invoke-static { p0 }, $INTEGRATIONS_CLASS_DESCRIPTOR->playerOverlayInflated(Landroid/view/ViewGroup;)V"
            )
        } ?: throw PlayerOverlaysOnFinishInflateFingerprint.exception
    }
}